### PR TITLE
Adding documentation

### PR DESCRIPTION
**Pull request by EmiCB**
- Updated the README
- Added JavaDocs and some inline comments to all files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,10 +1,97 @@
 # WHIMC-Portals
 [![GitHub release (latest by date)](https://img.shields.io/github/v/release/whimc/Portals?label=download&logo=github)](https://github.com/whimc/Portals/releases/latest)
 
-Create portals for player teleportation
+Portals is a Minecraft plugin that allows for the creation of custom portals with user-defined destinations. This plugin uses [Multiverse Portals](https://dev.bukkit.org/projects/multiverse-portals).
```
Comment (jackah2) [inline README.md]: We don't use multiverse portals - this plugin was meant to replace it actually.

File: README.md [hunk 2]
```diff
@@ -1,10 +1,97 @@
...
+/portal create <portal_name>
+/portal setfiller <portal_name> <block_type>
+/portal permission <permission_suffix>
+````
+
+After this, you'll want to configure a destination.
+
+### How to Create and Set a Destination
+
+Destinations are independent of portals, so you can have many portals link to a single destination. To create a destination to stand in the place you want it and do:
```
Comment (jackah2) [inline README.md]:
```suggestion
Destinations are independent of portals, so you can have many portals link to a single destination. To create a destination, stand where you'd like it to be and do:
```

File: README.md [hunk 3]
```diff
@@ -1,10 +1,97 @@
...
+|`/portal reshape <portal_name>`                      | Reshape a portal to your current selection.                        |
+|`/portal setfiller <portal_name> <block_type>`       | Sets the filler of a portal.                                       |
+|`/portal teleport <portal_name>`                     | Teleports the user to the provided portal.                         |
+|`/portal tool`                                       | Gives user the portal selector tool.                               |
+
+### How to Create a Portal
+
+To setup a new portal you'll want to use the `wooden sword` tool to `left` and then `right` click two blocks to select an ``empty`` area. Then do:
+
+```yaml
```
Comment (jackah2) [inline README.md]: These can all just be normal code blocks. You wouldn't want to format it as yaml unless the content is actually yaml

File: README.md [hunk 4]
```diff
@@ -1,10 +1,97 @@
...
+|`/destination create <destination_name>`             | Creates a new destination at the user's current position.          |
+|`/destination info <destination_name>`               | Displays information about the current destination.                |
+|`/destination list`                                  | Displays a list of all the existing destinations.                  |
+|`/destination purge <'invalid'/'no-portals'/'both'>` | Purges any unused and / or invalid destinations.                   |
+|`/destination remove <destination_name>`             | Removes a destination.                                             |
+|`/destination set <portal_name> <destination_name>`  | Sets the destination of a portal.                                  |
+|`/destination sethere <portal_name>`                 | Sets the destination of the portal to the user's current location. |
+|`/destination teleport <destination_name>`           | Teleports the user to the provided destination.                    |
+|`/portal create <portal_name>`                       | Creates a permissionless portal using the selected location.       |
```
Comment (jackah2) [inline README.md]:
```suggestion
|`/portal create <portal_name>`                       | Creates a permissionless portal using the selected area.       |
```

File: README.md [hunk 5]
```diff
@@ -1,10 +1,97 @@
...
+
+Note that `/destination` has many other sub-commands, including:
+
+![image](https://user-images.githubusercontent.com/5846359/136828198-ab093daf-1b08-481e-a352-f391a269c497.png)
+
+---
+## Configuration
+
+### portalData.yml
+
```
Comment (jackah2) [inline README.md:84]:
```suggestion

You shouldn't have to edit this file directly; everything can be done with commands.

```

File: README.md [hunk 6]
```diff
@@ -1,10 +1,97 @@
...
+|`/destination remove <destination_name>`             | Removes a destination.                                             |
+|`/destination set <portal_name> <destination_name>`  | Sets the destination of a portal.                                  |
+|`/destination sethere <portal_name>`                 | Sets the destination of the portal to the user's current location. |
+|`/destination teleport <destination_name>`           | Teleports the user to the provided destination.                    |
+|`/portal create <portal_name>`                       | Creates a permissionless portal using the selected location.       |
+|`/portal debug`                                      | Displays information about the portal being entered (no teleport). |
+|`/portal info <portal_name>`                         | Displays information about a portal.                               |
+|`/portal list`                                       | Lists all existing portals.                                        |
+|`/portal permission <permission_suffix>`             | Sets or removes portal permissions.                                |
```
Comment (jackah2) [inline README.md]: You should add that the required permission ends up being `whimc-portals.entry.<permission_suffix>`

File: README.md [hunk 7]
```diff
@@ -1,10 +1,97 @@
...
+## Commands
+
+| Command                                             | Description                                                        |
+|-----------------------------------------------------|--------------------------------------------------------------------|
+|`/destination change <destination_name>`             | Sets the location of a destination to the user's current position. |
+|`/destination clear <portal_name>`                   | Removes the destination of a portal.                               |
+|`/destination create <destination_name>`             | Creates a new destination at the user's current position.          |
+|`/destination info <destination_name>`               | Displays information about the current destination.                |
+|`/destination list`                                  | Displays a list of all the existing destinations.                  |
+|`/destination purge <'invalid'/'no-portals'/'both'>` | Purges any unused and / or invalid destinations.                   |
```
Comment (jackah2) [inline README.md]: Specify that "invalid" means it's in a non-existent world and "no-portals" just means it has no linked portals.

File: README.md [hunk 8]
```diff
@@ -1,10 +1,97 @@
...
+|`/destination set <portal_name> <destination_name>`  | Sets the destination of a portal.                                  |
+|`/destination sethere <portal_name>`                 | Sets the destination of the portal to the user's current location. |
+|`/destination teleport <destination_name>`           | Teleports the user to the provided destination.                    |
+|`/portal create <portal_name>`                       | Creates a permissionless portal using the selected location.       |
+|`/portal debug`                                      | Displays information about the portal being entered (no teleport). |
+|`/portal info <portal_name>`                         | Displays information about a portal.                               |
+|`/portal list`                                       | Lists all existing portals.                                        |
+|`/portal permission <permission_suffix>`             | Sets or removes portal permissions.                                |
+|`/portal purge <'invalid'/'no-destination'/'both'>`  | Purges any unused and / or invalid portals.                        |
```
Comment (jackah2) [inline README.md]: Same comment as the destination purge command

File: README.md [hunk 9]
```diff
@@ -1,10 +1,97 @@
...
+|`/portal refill <portal_name>`                       | Regenerates the filler of a portal.                                |
+|`/portal remove <portal_name>`                       | Removes a portal.                                                  |
+|`/portal reshape <portal_name>`                      | Reshape a portal to your current selection.                        |
+|`/portal setfiller <portal_name> <block_type>`       | Sets the filler of a portal.                                       |
+|`/portal teleport <portal_name>`                     | Teleports the user to the provided portal.                         |
+|`/portal tool`                                       | Gives user the portal selector tool.                               |
+
+### How to Create a Portal
+
+To setup a new portal you'll want to use the `wooden sword` tool to `left` and then `right` click two blocks to select an ``empty`` area. Then do:
```
Comment (jackah2) [inline README.md]:
```suggestion
To setup a new portal you'll want to do `/portal tool` to get the portal selection tool. Then left click to select your first position and right click to select your second. These two blocks will be the bounding corners of a cuboid like WorldEdit. Then do:
```

File: README.md [hunk 10]
```diff
@@ -1,10 +1,97 @@
...
+
+```yaml
+/portal create <portal_name>
+/portal setfiller <portal_name> <block_type>
+/portal permission <permission_suffix>
+````
+
```
Comment (jackah2) [inline README.md:57]:
```suggestion

Note that _only_ air blocks will be replaced with the filler block.

```

File: README.md [hunk 11]
```diff
@@ -1,10 +1,97 @@
...
+      z: 0
+    pos2:
+      x: 3
+      y: 3
+      z: 0
+    destination: DestinationName
+```
+
+---
+## Dependencies
```
Comment (jackah2) [inline README.md]: There are no dependencies

File: src/main/java/edu/whimc/portals/Destination.java [hunk 12]
```diff
@@ -47,10 +82,17 @@ private Destination(Main plugin, String name, Location location, String worldNam
 
     }
 
+    /** @return The list of destinations. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Destination.java]:
```suggestion
    /** @return The list of all destinations. */
```

File: src/main/java/edu/whimc/portals/Destination.java [hunk 13]
```diff
@@ -7,28 +7,63 @@
...
+ */
 public class Destination {
-
+    /* Keyword for no destination. */
     public static final String NONE = "none";
 
+    /* The instance of the plugin. */
     private Main plugin;
 
+    /* The list of destinations. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Destination.java]:
```suggestion
    /* The list of all destinations. */
```

File: src/main/java/edu/whimc/portals/Destination.java [hunk 14]
```diff
@@ -7,28 +7,63 @@
 import org.bukkit.Location;
 import org.bukkit.entity.Player;
 
+/**
+ * The class to store Destination data.
+ */
 public class Destination {
-
+    /* Keyword for no destination. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Destination.java]:
```suggestion
    /** Keyword for no destination. */
```
I believe one-liner javadocs like these should start with a double asterisk

File: src/main/java/edu/whimc/portals/Destination.java [hunk 15]
```diff
@@ -79,65 +128,90 @@ public static Destination getDestination(String name) {
         return linked;
     }
 
+    /** @return The name of the Destination. */
     public String getName() {
         return name;
     }
 
+    /** The location of the Destination. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Destination.java]:
```suggestion
    /** @return The location of the Destination. */
```

File: src/main/java/edu/whimc/portals/Portal.java [hunk 16]
```diff
@@ -23,37 +23,90 @@
 import org.bukkit.permissions.Permission;
 import org.bukkit.util.Vector;
 
+/**
+ * The class to sore Portal data.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Portal.java]:
```suggestion
 * The class to store Portal data.
```

File: src/main/java/edu/whimc/portals/Main.java [hunk 17]
```diff
@@ -18,14 +18,23 @@
...
     public static final String PERM_PREFIX = "whimc-portals";
 
+    /* The instance of the ConfigManager */
     private MyConfigManager manager;
+    /* The portals config. */
     private static MyConfig portalData;
+    /* The instance of the location saver. */
     private LocationSaver locationSaver;
 
+    /**
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Main.java]: You don't really need this javadoc

File: src/main/java/edu/whimc/portals/Portal.java [hunk 18]
```diff
@@ -23,37 +23,90 @@
...
+ */
 public class Portal {
 
+    /* The list of portals. */
     private static List<Portal> portals = new ArrayList<>();
+    /* The map of portal data. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Portal.java]:
```suggestion
    /* Map of serialized locations to their respective Portal. Allows for fast lookups. */
```

File: src/main/java/edu/whimc/portals/Portal.java [hunk 19]
```diff
@@ -23,37 +23,90 @@
 import org.bukkit.permissions.Permission;
 import org.bukkit.util.Vector;
 
+/**
+ * The class to sore Portal data.
+ */
 public class Portal {
 
+    /* The list of portals. */
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Portal.java]:
```suggestion
    /* The list of all portals. */
```
Same comment for the getPortals doc

File: src/main/java/edu/whimc/portals/Portal.java [hunk 20]
```diff
@@ -417,15 +625,23 @@ public boolean inPortal(Location location){
         return true;
     }
 
+    /**
+     * Sets the configuration of the portal.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/Portal.java]:
```suggestion
     * Sets a value in the config for this portal.
```

File: src/main/java/edu/whimc/portals/commands/destination/DestinationCommand.java [hunk 21]
```diff
@@ -55,6 +70,9 @@ public boolean onCommand(CommandSender sender, Command cmd, String commandLabel,
         return subCmd.executeSubCommand(sender, args);
     }
 
+    /**
+     * {@inheritDoc}
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/commands/destination/DestinationCommand.java:74]: Add a note about how this has the sub command handle this

File: src/main/java/edu/whimc/portals/commands/portal/PortalDebug.java [hunk 22]
```diff
@@ -9,14 +9,31 @@
 import edu.whimc.portals.utils.Messenger;
 import edu.whimc.portals.utils.Messenger.Message;
 
+/**
+ * The command to toggle debug more. Displays information about the portal the user is entering
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/commands/portal/PortalDebug.java]:
```suggestion
 * The command to toggle debug mode for the player. Displays information about the portal the user is entering
```

File: src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java [hunk 23]
```diff
@@ -10,26 +10,49 @@
...
 
-public class PortalBlockChangeListener implements Listener{
-
+/**
+ * Listens for events where the portal blocks change.
+ */
+public class PortalBlockChangeListener implements Listener {
+
+    /**
+     * Cancel the water movement if the portal exists.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java]:
```suggestion
     * Cancel water/lava movement inside portals.
```

File: src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java [hunk 24]
```diff
@@ -10,26 +10,49 @@
...
+     * @param event The block event.
+     */
     @EventHandler(priority = EventPriority.HIGH)
     public void onWaterMove(BlockFromToEvent event){
         Portal portal = Portal.getPortal(event.getBlock());
         if (portal != null) event.setCancelled(true);
     }
 
+    /**
+     * Cancel the block formation if the portal exists.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java]:
```suggestion
     * Cancel block formations inside portals.
```

File: src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java [hunk 25]
```diff
@@ -10,26 +10,49 @@
...
+     * @param event The block event.
+     */
     @EventHandler(priority = EventPriority.HIGH)
     public void onBlockForm(BlockFormEvent event) {
         Portal portal = Portal.getPortal(event.getBlock());
         if (portal != null) event.setCancelled(true);
     }
 
+    /**
+     * Cancel the block formation if the portal exists.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java]:
```suggestion
     * Stop entities from forming blocks inside portals.
```

File: src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java [hunk 26]
```diff
@@ -10,26 +10,49 @@
...
+     * @param event The block event.
+     */
     @EventHandler(priority = EventPriority.HIGH)
     public void onEntityBlockForm(EntityBlockFormEvent event) {
         Portal portal = Portal.getPortal(event.getBlock());
         if (portal != null) event.setCancelled(true);
     }
 
+    /**
+     * Cancel the block ignition if the portal exists.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java]:
```suggestion
     * Cancel block ignition inside portals.
```

File: src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java [hunk 27]
```diff
@@ -10,26 +10,49 @@
 
 import edu.whimc.portals.Portal;
 
-public class PortalBlockChangeListener implements Listener{
-
+/**
+ * Listens for events where the portal blocks change.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalBlockChangeListener.java]:
```suggestion
 * Listens for events where the blocks inside portals change.
```

File: src/main/java/edu/whimc/portals/listeners/PortalDamageListener.java [hunk 28]
```diff
@@ -9,21 +9,41 @@
 
 import edu.whimc.portals.Portal;
 
+/**
+ * Listens for events that cause damage to the portal.
+ */
 public class PortalDamageListener implements Listener {
 
+    /**
+     * Cancel the combustion event if it is too close to a portal.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalDamageListener.java]:
```suggestion
     * Stop entities from lighting on fire inside portals. (For lava-filled portals).
```

File: src/main/java/edu/whimc/portals/listeners/PortalDamageListener.java [hunk 29]
```diff
@@ -9,21 +9,41 @@
...
+     * @param event The EntityCombustByBlockEvent.
+     */
     @EventHandler
     public void onCombust(EntityCombustByBlockEvent event) {
         Portal near = getNearPortal(event.getEntity().getLocation().getBlock(), 1);
         if (near != null) event.setCancelled(true);
     }
 
+    /**
+     * Cancel damage caused to entities if it is too close to a portal.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalDamageListener.java]:
```suggestion
     * Stop entitles from taking lava damage from lava-filled portals.
```

File: src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java [hunk 30]
```diff
@@ -19,66 +19,105 @@
...
-
+/**
+ * Listens for events where the portal is entered.
+ */
+public class PortalEnterListener implements Listener {
+    /* The list of UUIDs of players with debug mode enabled. */
     private static Set<UUID> debugPlayers = new HashSet<>();
 
+    /**
+     * Adds the UUID of the passed player to hte list of players with debug mode enabled.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java]:
```suggestion
     * Enable debug mode for the given player.
```
You don't want to get too into the specifics of the implementation.

File: src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java [hunk 31]
```diff
@@ -19,66 +19,105 @@
...
+     *
+     * @param player The player.
+     */
     public static void addDebugPlayer(Player player) {
         debugPlayers.add(player.getUniqueId());
     }
 
+    /**
+     * Removes the UUID of the passed player to the list of players with debug mode enabled.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java]:
```suggestion
     * Disable debug mode for the given player.
```

File: src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java [hunk 32]
```diff
@@ -19,66 +19,105 @@
...
-    public void onMove(PlayerMoveEvent event){
+    public void onMove(PlayerMoveEvent event) {
+        // get the move event locations and check if it's withing the portal bounds
         Location locTo = event.getTo();
         Location locFrom = event.getFrom();
         if (locFrom.getBlockX() == locTo.getBlockX() && locFrom.getBlockZ() == locTo.getBlockZ() && locFrom.getBlockY() == locTo.getBlockY()) return;
 
         Portal portal = Portal.getPortal(locTo);
-        if(portal == null) return;
+        if (portal == null) return;
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java:70]: Nice catch thank you

File: src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java [hunk 33]
```diff
@@ -19,66 +19,105 @@
...
 
+    /**
+     * Send info / teleport if player moves into a portal.
+     *
+     * @param event The PlayerMoveEvent.
+     */
     @EventHandler(priority = EventPriority.HIGHEST)
-    public void onMove(PlayerMoveEvent event){
+    public void onMove(PlayerMoveEvent event) {
+        // get the move event locations and check if it's withing the portal bounds
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/PortalEnterListener.java]:
```suggestion
        // return early if the player didn't move a full block.
```

File: src/main/java/edu/whimc/portals/listeners/ToolSelectListener.java [hunk 34]
```diff
@@ -15,40 +15,64 @@
 
 import edu.whimc.portals.utils.Messenger;
 
-public class ToolSelectListener implements Listener{
-
+/**
+ * Listens for events where a portal tool is selected.
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/listeners/ToolSelectListener.java]:
```suggestion
 * Listens for events where a portal tool is used.
```

File: src/main/java/edu/whimc/portals/utils/Messenger.java [hunk 35]
```diff
@@ -9,42 +9,102 @@
...
+     * @param description The command's description.
+     * @param usage The command's usage.
+     */
     public static void usageMissingArgs(CommandSender sender, String missingArgs, String description, String usage) {
         msg(sender, ReplaceMessage.MISSING_ARGUMENTS, missingArgs);
         msg(sender, "  " + usage);
     }
 
+    /**
+     * Sends the sender a message about what argument are is unknown and displays the command's
```
Comment (jackah2) [inline src/main/java/edu/whimc/portals/utils/Messenger.java]:
```suggestion
     * Sends the sender a message about providing an unknown argument and displays the command's
```

**Review by jackah2 — CHANGES_REQUESTED** [review]
A handful of changes but pretty minor ones. Thank you for doing this and catching a lot of my mistakes :)

File: README.md (hunk 3 above)
Comment (EmiCB) [reply README.md]: Whoops! Looks like I just copy pasted them and forgot to remove the language specifier lol

**Review by jackah2 — APPROVED** [review]
Looks good! Thank you for doing this!

**Pull request merged**